File: javascript/radixSort.js
function countingSort(arr, exp) {
  const n = arr.length;
  const output = new Array(n);
  const count = new Array(10).fill(0);

  // store count of occurrences in coutn[]
  for(let i = 0; i < n; i++) {
    const index = Math.floor(arr[i] / exp) % 10;
    count[index]++;
  }

  for(let i = 1; i < 10; i++) {
    count[i] += count[i - 1];
  }
  // build output array
  for(let i = n - 1; i >= 0; i--) {
    const index = Math.floor(arr[i] / exp) % 10;
    output[count[index] - 1] = arr[i];
    count[index]--;
  }

  for(let i = 0; i < n; i++) {
    arr[i] = output[i];
  }
}
function radixSort(arr) {
  const max = Math.max(arr);
  let exp = 1;
  while(Math.floor(max / exp) > 0) {
    countingSort(arr, exp);
    exp*=10;
  }
}
// example usage
const arr = [170, 45, 75, 90, 802, 24, 2, 66];
radixSort(arr);
console.log(arr);

File: python/radix_sort.py
def countingSort(arr, exp1):
  n = len(arr)
  output = [0] * n
  count = [0] * 10

  for i in range(0, n):
    index = (arr[i] / exp1) % 10
    count[index] += 1

  for i in range(1, 10):
    count[i] += count[i - 1]

  i = n - 1
  while i >= 0:
    index = (arr[i] / exp1) % 10
    output[count[index] - 1] = arr[i]
    count[index] -= 1
    i -= 1

  i = 0
  for i in range(0, n):
    arr[i] = output[i]

def radixSort(arr):
  max1 = max(arr)
  exp = 1
  while max1 / exp > 0:
    countingSort(arr, exp)
    exp*= 10

# example usage
arr = [170, 45, 75, 90, 802, 24, 2, 66]
radixSort(arr)
print(arr) 
